itests/eth_filter_test.go

```diff
@@ -622,6 +637,123 @@ func TestTxReceiptBloom(t *testing.T) {
+
+	// DEPLOY CONTRACT
+	tx := deployContract(ctx, t, client, ethAddr, "./contracts/MultipleEvents.hex")
+
+	client.EVM().SignTransaction(tx, key.PrivateKey)
+	hash := client.EVM().SubmitTransaction(ctx, tx)
+
+	receipt, err := client.EVM().WaitTransaction(ctx, hash)
+	require.NoError(t, err)
+	require.NotNil(t, receipt)
+	require.EqualValues(t, ethtypes.EthUint64(0x1), receipt.Status)
+
+	// invoke function to emit four events
+	contractAddr := client.EVM().ComputeContractAddress(ethAddr, 0)
+
+	// https://abi.hashex.org/
+	params4Events, err := hex.DecodeString("98e8da00")
+	require.NoError(t, err)
+
+	params3Events, err := hex.DecodeString("05734db70000000000000000000000001cd1eeecac00fe01f9d11803e48c15c478fe1d22000000000000000000000000000000000000000000000000000000000000000b")
+	require.NoError(t, err)
+
+	gasParams, err := json.Marshal(ethtypes.EthEstimateGasParams{Tx: ethtypes.EthCall{
+		From: &ethAddr,
+		To:   &contractAddr,
+		Data: params4Events,
+	}})
+	require.NoError(t, err)
+
+	gaslimit, err := client.EthEstimateGas(ctx, gasParams)
+	require.NoError(t, err)
+
+	maxPriorityFeePerGas, err := client.EthMaxPriorityFeePerGas(ctx)
+	require.NoError(t, err)
+
+	paramsArray := [][]byte{params4Events, params3Events, params3Events}
+	var receipts []*api.EthTxReceipt
+	var hashes []ethtypes.EthHash
+
+	for i, params := range paramsArray {
+		invokeTx := ethtypes.Eth1559TxArgs{
+			ChainID:              build.Eip155ChainId,
+			To:                   &contractAddr,
+			Value:                big.Zero(),
+			Nonce:                i + 1,
+			MaxFeePerGas:         types.NanoFil,
+			MaxPriorityFeePerGas: big.Int(maxPriorityFeePerGas),
+			GasLimit:             int(gaslimit),
+			Input:                params,
+			V:                    big.Zero(),
+			R:                    big.Zero(),
+			S:                    big.Zero(),
+		}
+
+		// Submit transaction with valid signature
+		client.EVM().SignTransaction(&invokeTx, key.PrivateKey)
+		hash := client.EVM().SubmitTransaction(ctx, &invokeTx)
+		hashes = append(hashes, hash)
+	}
+
+	for _, hash := range hashes {
+		receipt, err := client.EVM().WaitTransaction(ctx, hash)
+		require.NoError(t, err)
+		require.NotNil(t, receipt)
+		receipts = append(receipts, receipt)
+	}
+
+	// Assert that first receipt has 4 event logs and the other two have 3 each
+	require.Equal(t, 4, len(receipts[0].Logs), "First transaction should have 4 event logs")
+	require.Equal(t, 3, len(receipts[1].Logs), "Second transaction should have 3 event logs")
+	require.Equal(t, 3, len(receipts[2].Logs), "Third transaction should have 3 event logs")
+
+	// Iterate over all logs in all receipts and assert that the logIndex field is numbered from 0 to 9
+	expectedLogIndex := 0
+	var receiptLogs []ethtypes.EthLog
+	for _, receipt := range receipts {
+		for _, log := range receipt.Logs {
+			require.Equal(t, ethtypes.EthUint64(expectedLogIndex), log.LogIndex, "LogIndex should be sequential")
+			expectedLogIndex++
```

Comment:
This assumes that the messages are going to land in the same tipset; I'm not sure that's a safe assumption. You have a slower blocktime here (1s) which helps, but you're still just throwing 3 at the mpool and assuming they get picked up at the same time, what if they're not?

Some ideas:
* Add more than 3, to increase probability of >1 landing in a single tipset; then group receipts by tipset and loop through tipsets and only count expectedLogIndex _per tipset_
* Put this whole block in a loop, (give it ~3 retries? more?) such that if you check the receipts and find that they are all in different tipsets then you bork and repeat the loop until they do, then check that they line up _per tipset_

Reply:
Done.